tests/block_test_util.go

```diff
@@ -194,6 +238,108 @@ func (t *BlockTest) insertBlocks(blockchain *blockchain.BlockChain) ([]btBlock,
+
+	// insert the test blocks, which will execute all transactions
+	for _, b := range t.json.Blocks {
+		txs, header, err := b.decodeTx()
+		if err != nil {
+			if b.BlockHeader == nil {
+				continue // OK - block is supposed to be invalid, continue with next block
+			} else {
+				return nil, nil, nil, fmt.Errorf("Block RLP decoding failed when expected to succeed: %v", err)
+			}
+		}
+		// RLP decoding worked, try to insert into chain:
+		kaiaReward := common.Big0
+		ethReward := common.Big0
+
+		// The intrinsic gas calculation affects gas used, so we need to make some changes to the main code.
+		if bc.Config().IsIstanbulForkEnabled(bc.CurrentHeader().Number) {
+			types.IsPragueInExecutionSpecTest = true
+		}
+		blockchain.GasLimitInExecutionSpecTest = header.GasLimit
+
+		// var maxFeePerGas *big.Int
+		blocks, receiptsList := blockchain.GenerateChain(bc.Config(), preBlock, bc.Engine(), db, 1, func(i int, b *blockchain.BlockGen) {
+			b.SetRewardbase(common.Address(header.Coinbase))
+			for _, tx := range txs {
+				b.AddTxWithChainEvenHasError(bc, tx)
+			}
+		})
+		preBlock = blocks[0]
+
+		// The reward calculation is different for kaia and eth, and this will be deducted from the state later.
+		for _, receipt := range receiptsList[0] {
+			for _, tx := range blocks[0].Body().Transactions {
+				if tx.Hash() != receipt.TxHash {
+					continue
+				}
+
+				// kaia gas price
+				var kaiaGasPrice *big.Int
+				if tx.Type() == types.TxTypeEthereumDynamicFee || tx.Type() == types.TxTypeEthereumSetCode {
+					kaiaGasPrice = tx.EffectiveGasPrice(blocks[0].Header(), bc.Config())
+				} else {
+					kaiaGasPrice = tx.GasPrice()
+				}
+
+				// eth gas price
+				ethGasPrice := tx.GasPrice()
+				if header.BaseFee != nil {
+					ethGasPrice = math.BigMin(new(big.Int).Add(tx.GasTipCap(), header.BaseFee), tx.GasFeeCap())
+				}
+
+				// Record kaia's reward.
+				kaiaReward = new(big.Int).Add(kaiaReward, new(big.Int).Mul(new(big.Int).SetUint64(receipt.GasUsed), kaiaGasPrice))
+
+				// Record eth's reward.
+				ethReward = new(big.Int).Add(ethReward, calculateEthMiningReward(ethGasPrice, tx.GasFeeCap(), tx.GasTipCap(), header.BaseFee,
+					receipt.GasUsed, bc.Config().Rules(blocks[0].Header().Number)))
+
+				// Because it is a eth test, we don't have to think about fee payer
+				// Because the baseFee is set to 0, Kaia's gas fee may be 0 if the transaction has a dynamic fee.
+				senderMap[tx.ValidatedSender()] = new(big.Int).Sub(
+					new(big.Int).Mul(new(big.Int).SetUint64(receipt.GasUsed), kaiaGasPrice),
+					new(big.Int).Mul(new(big.Int).SetUint64(receipt.GasUsed), ethGasPrice))
+			}
+		}
+
+		if header.GasUsed != blocks[0].GasUsed() {
+			return nil, nil, nil, fmt.Errorf("Unexpected GasUsed error (Expected: %v, Actual: %v)", header.GasUsed, blocks[0].GasUsed())
+		}
+
+		rewardMap[common.Address(header.Coinbase)] = rewardList{
+			kaiaReward: kaiaReward,
+			ethReward:  ethReward,
+		}
+
+		i, err := bc.InsertChain(blocks)
```

Comment:
Is the same block executed again?

Reply:
No, only one block is generated by GenerateChain, so one block is inserted.